velox/exec/Window.cpp

```diff
@@ -0,0 +1,475 @@
+  // the rows.
+  sortedRows_.resize(numRows_);
+  RowContainerIterator iter;
+  data_->listRows(&iter, numRows_, sortedRows_.data());
+
+  std::sort(
+      sortedRows_.begin(),
+      sortedRows_.end(),
+      [this](const char* leftRow, const char* rightRow) {
+        return compareRowsWithKeys(leftRow, rightRow, allKeyInfo_);
+      });
+
+  computePartitionStartRows();
+
+  currentPartition_ = 0;
+}
+
+void Window::noMoreInput() {
+  Operator::noMoreInput();
+  // No data.
+  if (numRows_ == 0) {
+    finished_ = true;
+    return;
+  }
+
+  // At this point we have seen all the input rows. We can start
+  // outputting rows now.
+  // However, some preparation is needed. The rows should be
+  // separated into partitions and sort by ORDER BY keys within
+  // the partition. This will order the rows for getOutput().
+  sortPartitions();
+  createPeerAndFrameBuffers();
+}
+
+std::pair<vector_size_t, vector_size_t> Window::findFrameEndPoints(
+    vector_size_t /*functionNumber*/,
+    vector_size_t partitionStartRow,
+    vector_size_t /*partitionEndRow*/,
+    vector_size_t currentRow) {
+  // TODO : We handle only the default window frame in this code. Add support
+  // for all window frames subsequently.
+
+  // Default window frame is Range UNBOUNDED PRECEDING CURRENT ROW.
+  return std::make_pair(partitionStartRow, currentRow);
+}
+
+void Window::callResetPartition(vector_size_t partitionNumber) {
+  auto partitionSize = partitionStartRows_[partitionNumber + 1] -
+      partitionStartRows_[partitionNumber];
+  auto partition = folly::Range(
+      sortedRows_.data() + partitionStartRows_[partitionNumber], partitionSize);
+  windowPartition_->resetPartition(partition);
+  for (int i = 0; i < windowFunctions_.size(); i++) {
+    windowFunctions_[i]->resetPartition(windowPartition_.get());
+  }
+}
+
+// Call WindowFunction::apply for the rows between startRow and endRow in
+// the sortedRows_ ordering of the data_ RowContainer.
+void Window::callApplyForPartitionRows(
+    vector_size_t startRow,
+    vector_size_t endRow,
+    const std::vector<VectorPtr>& result,
+    vector_size_t resultOffset) {
+  if (partitionStartRows_[currentPartition_] == startRow) {
+    callResetPartition(currentPartition_);
+  }
+
+  auto peerCompare = [&](const char* lhs, const char* rhs) -> bool {
+    return compareRowsWithKeys(lhs, rhs, sortKeyInfo_);
+  };
+
+  vector_size_t numRows = endRow - startRow;
+  vector_size_t numFuncs = windowFunctions_.size();
+
+  // Size buffers for the call to WindowFunction::apply.
+  auto bufferSize = numRows * sizeof(vector_size_t);
+  peerStartBuffer_->setSize(bufferSize);
+  peerEndBuffer_->setSize(bufferSize);
+  for (auto w = 0; w < numFuncs; w++) {
+    frameStartBuffers_[w]->setSize(bufferSize);
+    frameEndBuffers_[w]->setSize(bufferSize);
+  }
+
+  // Setup values in the peer and frame buffers.
+  auto firstPartitionRow = partitionStartRows_[currentPartition_];
```

Comment:
Do we have test coverage for the calculation of peer start / end? It looks like we might not have it. If that's the case, then I suggest replacing this code with a TODO and adding it back along with usage and test coverage or adding window functions to this PR which would use this information and adding tests for the same.

Reply:
I don't believe we have test coverage for this code. Let's remove it and add it back along with tests in future PR.